import "goby/common/option_extensions.proto";

package goby.acomms.protobuf;

message DriverConfig
{
  optional int32 modem_id = 1 [(goby.field).description="Unique number 1-31 to identify this node", (goby.field).example="1"];
  
  enum ConnectionType { CONNECTION_SERIAL = 1; // Modem is connected by a tty serial line (e.g. RS-232)
                        CONNECTION_TCP_AS_CLIENT = 2; // Modem is connected by ethernet and is serving clients
                        CONNECTION_TCP_AS_SERVER = 3; // Modem is connected by ethernet and expects us to serve it
                        CONNECTION_DUAL_UDP_BROADCAST = 4; // Presently unimplemented
  }
  
  optional ConnectionType connection_type = 2 [default = CONNECTION_SERIAL, (goby.field).description="Physical connection type from this computer (running Goby) to the acoustic modem"];
  optional string line_delimiter = 3 [default = "\r\n", (goby.field).description="String used to delimit new lines for this acoustic modem"]; // delimits lines on the serial or tcp interface
  
  // required if connection_type == CONNECTION_SERIAL
  optional string serial_port = 4 [(goby.field).description="Serial port for CONNECTION_SERIAL", (goby.field).example="/dev/ttyS0"]; 
  optional uint32 serial_baud = 5 [(goby.field).description="Baud rate for CONNECTION_SERIAL", (goby.field).example="19200"]; 

  // required if connection_type == CONNECTION_TCP_AS_CLIENT
  optional string tcp_server = 6 [(goby.field).description="IP Address or domain name for the server if CONNECTION_TCP_AS_CLIENT", (goby.field).example="192.168.1.111"]; 

  // required if connection_type == CONNECTION_TCP_AS_CLIENT or CONNECTION_TCP_AS_SERVER
  optional uint32 tcp_port = 7 [(goby.field).description="Port to serve on (for CONNECTION_TCP_AS_SERVER) or to connect to (for CONNECTION_TCP_AS_CLIENT)", (goby.field).example="50010"];
  
  
  extensions 1000 to max;
  // extensions 1000-1100 used by acomms_mm_driver.proto
  // exenstions 1201-1202 used by acomms_abc_driver.proto (example driver)
}

